src/daft-csv/src/read.rs

```diff
@@ -199,6 +220,387 @@ fn tables_concat(mut tables: Vec<Table>) -> DaftResult<Table> {
+        }
+    }
+    buf.pool.return_buffer(chunk_buffer);
+    Ok(tables)
+}
+
+async fn stream_csv_local(
+    uri: &str,
+    convert_options: Option<CsvConvertOptions>,
+    parse_options: CsvParseOptions,
+    read_options: Option<CsvReadOptions>,
+    max_chunks_in_flight: Option<usize>,
+) -> DaftResult<impl Stream<Item = DaftResult<Table>> + Send> {
+    let uri = uri.trim_start_matches("file://");
+    let file = std::fs::File::open(uri)?;
+    let mmap = unsafe { memmap2::Mmap::map(&file) }.context(StdIOSnafu)?;
+    let bytes = &mmap[..];
+
+    // TODO(desmond): This logic is repeated multiple times in this file. Should dedup.
+    let predicate = convert_options
+        .as_ref()
+        .and_then(|opts| opts.predicate.clone());
+
+    let limit = convert_options.as_ref().and_then(|opts| opts.limit);
+
+    let include_columns = convert_options
+        .as_ref()
+        .and_then(|opts| opts.include_columns.clone());
+
+    let convert_options = match (convert_options, &predicate) {
+        (None, _) => None,
+        (co, None) => co,
+        (Some(mut co), Some(predicate)) => {
+            if let Some(ref mut include_columns) = co.include_columns {
+                let required_columns_for_predicate = get_required_columns(predicate);
+                for rc in required_columns_for_predicate {
+                    if include_columns.iter().all(|c| c.as_str() != rc.as_str()) {
+                        include_columns.push(rc)
+                    }
+                }
+            }
+            // If we have a limit and a predicate, remove limit for stream.
+            co.limit = None;
+            Some(co)
+        }
+    }
+    .unwrap_or_default();
+    // End of `should dedup`.
+
+    // TODO(desmond): We should do better schema inference here.
+    let schema = convert_options.clone().schema.unwrap().to_arrow()?;
+    let n_threads: usize = std::thread::available_parallelism()
+        .unwrap_or(NonZeroUsize::new(2).unwrap())
+        .into();
+    let chunk_size = read_options
+        .as_ref()
+        .and_then(|opt| opt.chunk_size.or_else(|| opt.buffer_size.map(|bs| bs / 8)))
+        .unwrap_or(DEFAULT_CHUNK_SIZE);
+    let projection_indices = fields_to_projection_indices(
+        &schema.clone().fields,
+        &convert_options.clone().include_columns,
+    );
+    let fields = schema.clone().fields;
+    let fields_subset = projection_indices
+        .iter()
+        .map(|i| fields.get(*i).unwrap().into())
+        .collect::<Vec<daft_core::datatypes::Field>>();
+    let read_schema = Arc::new(daft_core::schema::Schema::new(fields_subset)?);
+    let read_daft_fields = Arc::new(
+        read_schema
+            .fields
+            .values()
+            .map(|f| Arc::new(f.clone()))
+            .collect::<Vec<_>>(),
+    );
+    // TODO(desmond): Need better upfront estimators. Sample or keep running count of stats.
+    let estimated_mean_row_size = 100f64;
+    let estimated_std_row_size = 20f64;
```

Comment:
you should be able to do something similar to what I did in JSON [here](https://github.com/universalmind303/Daft/blob/7b40f23a5ff83aba4ab059b62ac781d7766be0b1/src/daft-json/src/local.rs#L338) to get the estimated sizes